### PR TITLE
Introduce a Union-Find structure to make constraints' code cleaner

### PR DESCRIPTION
**Pull request by Niols**
This PR builds on top of #146. It comes from the remark that the code for constraints unification basically inlines the implementation of a union-find data structure, which makes it harder to read, harder to optimise (with eg. the typical union-find optimisations) and more error-prone. The current PR implements a tiny union-find module and 

## What is a Union-Find data structure?

Not going to dig into history here, but a union-find data structure basically allows to represent efficiently equivalence classes over values as long as there exists a total comparison function. A Union-Find has two main functions:

- `union` merges two equivalence classes together. This typically has the type `key -> key -> UnionFind key -> UnionFind key`.

- `find` returns a representative of the given equivalence class. This typically has the type `key -> UnionFind key` with the invariant that for all `uf`, `k1`, `k2`, `find k1 uf == find k2 uf` if and only if `k1` and `k2` are in the same equivalence class. This function is typically not exposed to the user to which we rather expose an equivalence test of type `key -> key -> UnionFind key -> Bool`.

Union-Find can very easily be extended to represent a map from equivalence classes to values. Such a structure has three main functions:

- `union` merges together two equivalence classes. Because these equivalence classes are associated to values, one must specify how to merge the values as well. Typically, the type would be `union :: (value -> value -> value) -> key -> key -> UnionFind key value -> UnionFind key value`.

- `find` returns a representative of the given equivalence class as well as the associated value. This typically has the type `key -> UnionFind key value -> (key, Maybe value)`. Once again, this function is typically not exposed and we will prefer to expose an equivalence test (`key -> key -> UnionFind key value -> Bool`) and a lookup (`key -> UnionFind key value -> Maybe value`).

- `insert` (or `add` or whatever, I don't think there is a well-defined name for that) adds a mapping from key to value. Because there might already be a mapping from this equivalence class to a value, one must specify how to merge the values. Typically, the type would be `insert :: (value -> value -> value) -> key -> value -> UnionFind key value -> UnionFind key value`.

Some optimisations can make such structures very efficient with quasi-linear time for both operations. One of them requires that the `find` operation itself updates the structure, giving the types:

```haskell
union :: (value -> value -> value) -> key -> key -> UnionFind key value -> UnionFind key value
--       ^^^ unchanged

find :: key -> UnionFind key value -> (UnionFind key value, key, Maybe value)
areEquivalent :: key -> key -> UnionFind key value -> (UnionFind key value, Bool)
lookup :: key -> UnionFind key value -> (UnionFind key value, Maybe value)

insert :: (value -> value -> value) -> key -> value -> UnionFind key value -> UnionFind key value
--        ^^^ unchanged
```

(Note: a state monad would make those types much nicer.)

## Where is this inlined Union-Find?

As of f12b951 on #146, a set of constraints is represented as follows:

https://github.com/tweag/pirouette/blob/f12b95104b8b17996228587867c85f193944b7ab/src/Pirouette/SMT/Constraints.hs#L29-L41

Note in particular the fields `csAssignments :: M.Map meta (TermMeta lang meta)` and `csMetaEq :: M.Map meta meta` that only make sense when taken together:

- `csMetaEq` maps meta-variables to other meta-variables with the invariant that if `v` maps to `v'`, then `v'` is smaller than `v` for some comparison (here, the lexicographic order, but it does not matter). `csMetaEq` therefore represents an equivalence relation of meta-variables as well as a way to pick their representative: the smallest element according to the comparison.

- `csAssignments` maps meta-variables to terms with the invariant (not mentioned in the comments) that a key in `csAssignments` is always a representative in `csMetaEq`. It is therefore intended as a map from equivalence classes to terms.

Keeping those invariants brings complexity to the unification functions. See for instance `unifyMetaWith`:

https://github.com/tweag/pirouette/blob/f12b95104b8b17996228587867c85f193944b7ab/src/Pirouette/SMT/Constraints.hs#L135-L147

and in particular line 142 that resolves the representative of the equivalence class of the given meta-variable. In this example, abstracting and using a union-find library would bring two advantages. First, it would make the code simpler by removing some cases. Second, it would make clearer what happens in some corner cases. For instance, in `unifyMetaWith`, what happens if there is an equivalence class in `csMetaEq` but no binding for it in `csAssignments`? The variable is considered to be a *new* meta-variable and is sent to `unifyNewMetaWith`. Is this a bug? Is this intended? Because it sure seem to be violating `unifyNewMetaWith`'s invariant.

## How would it look with a Union-Find?

Look at the code! In 178003f21fbbb5d5aeb7a9318a88bcdc9af81d2f, there is in particular:

https://github.com/tweag/pirouette/blob/178003f21fbbb5d5aeb7a9318a88bcdc9af81d2f/src/Pirouette/SMT/Constraints.hs#L30-L40

and

https://github.com/tweag/pirouette/blob/178003f21fbbb5d5aeb7a9318a88bcdc9af81d2f/src/Pirouette/SMT/Constraints.hs#L130-L144

## Additional Notes and Questions

1. Is it possible for meta-variables to be equivalent (and mapped in `csMetaEq` as such) without having any binding in `csAssignments`? I tried to assume that it wasn't but the “OHearn” test (and only that one) started to fail. It goes against my initial intuition but I see now why it would make sense. I would therefore need someone's opinion on this.

1. Some previous functionalities are missing (eg. the `Monoid` instance for `ConstraintSet`; actually, that one seems buggy to me in the previous implementation). They do not seem used anywhere, though?

1. At this point, I did not implement any of the optimisations of the union-find data structure. The union-find's functions' interface already prepares for this, though.

1. Maybe we want to give a better interface to this union-find module by using a state monad somewhere?

1. A persistent union-find is a peculiar data structure in that the function `find` modifies the structure *without changing its semantics*. In this particular case, it means that it is possible to make `find` actually mutate the memory without violating the interesting properties of persistence. This gives an extra tiny speedup in case the union-find is used in several places without much differences. This makes things a bit dirty though.

### CODE REVIEW

File: src/Pirouette/SMT/Constraints.hs [hunk 1]
```diff
@@ -25,50 +25,45 @@ import Pirouette.Term.Syntax
...
-  s1 <> s2 =
-    ConstraintSet
-      (csAssignments s1 <> csAssignments s2)
-      (csMetaEq s1 <> csMetaEq s2)
-      (csNative s1 <> csNative s2)
-      (csRelations s1 <> csRelations s2)
+-- instance (Ord meta, Semigroup meta) => Semigroup (ConstraintSet lang meta) where
+--   s1 <> s2 =
+--     ConstraintSet
+--       (csAssignments s1 <> csAssignments s2)
```
Comment (Niols) [inline src/Pirouette/SMT/Constraints.hs:45]: This would need to be implemented for Union-Find, probably by requiring the values (here, `TermMeta lang meta`) to be a semigroup as well. Otherwise, we end up with something potentially buggy, as I think is the case in #146 (cf https://github.com/tweag/pirouette/pull/146#discussion_r964637039).

File: src/Pirouette/SMT/Constraints.hs [hunk 2]
```diff
@@ -153,23 +150,16 @@ conjunct c cs0 =
...
+      cs { csAssignments = UF.trivialUnion v u (csAssignments cs) }
     unifyNewMetaWith cs v u =
-      cs {csAssignments = M.insert v u (csAssignments cs)}
+      cs { csAssignments = UF.trivialInsert v u (csAssignments cs) }
+    -- FIXME: The fact that we are only using trivial unions and insert probably
+    -- means that we are still doing the job of the union-find. At least, we
+    -- have it to back us up and check that these insertions and unions are
+    -- indeed trivial. We should instead define what needs to happen when unions
+    -- are non-trivial, and then we could get rid of the difference between
+    -- `unifyMetaWith` and `unifyNewMetaWith`.
```
Comment (Niols) [inline src/Pirouette/SMT/Constraints.hs:162]: !! I will try to fix this tomorrow. It would really make the code much nicer.

File: src/UnionFind.hs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            let (unionFind', ancestor, _) = find key unionFind in
+              (unionFind', (key, ancestor) : equalities, bindings)
+          Ancestor ->
+            (unionFind, equalities, bindings)
+          AncestorWith value ->
+            (unionFind, equalities, (key, value) : bindings))
+    (unionFind, [], [])
+    (Map.toList $ getMap unionFind)
+
+join :: Maybe (Maybe value) -> Maybe value
```
Comment (Niols) [inline src/UnionFind.hs:198]: This is just a helper (that would actually work for any monad) but I don't think it's present in any library that we are using? And it feels a bit dumb to add a dependency just for this.

File: src/UnionFind.hs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+-- FIXME: Implement the optimisation for future calls
+--
+find :: Ord key =>
+  key ->
+  UnionFind key value ->
+  (UnionFind key value, key, Maybe (Maybe value))
+find key unionFind =
+  case Map.lookup key $ getMap unionFind of
+    Nothing -> (unionFind, key, Nothing)
+    Just (ChildOf key') -> find key' unionFind -- will be @Just@
```
Comment (Niols) [inline src/UnionFind.hs:80]: Should we split `find` into two functions, adding one where we know for sure that the result exists (and `error` otherwise) and that we would call from here to enforce the invariant?

**Comment by aspiwack** [comment]
For the record (understanding that it's easy to change in the future), in my experience, a mutable implementation (based on `ST`) is quite a bit faster. So if performance is an issue it's something to think about.

File: src/Pirouette/SMT/Constraints.hs [hunk 5]
```diff
@@ -25,50 +25,45 @@ import Pirouette.Term.Syntax
...
-      prettyNative n = pretty (show n)
+instance (Pretty meta, Pretty (TermMeta lang meta)) => Pretty (ConstraintSet lang meta) where
+  pretty ConstraintSet {..} = undefined
+    -- vsep $
+    --   [pretty csAssignments]
+    --     ++ map prettyRelations csRelations
+    --     ++ map prettyNative csNative
+    -- where
+    --   prettyRelations (r, v, u) = pretty v <+> pretty r <+> pretty u
+    --   prettyNative n = pretty (show n)
```
[the remark this answers is not in the document]
Comment (lucaspena) [reply src/Pirouette/SMT/Constraints.hs]: I was able to get it to work by basically copying the pretty code in the `UnionFind` module. Otherwise, I was only able to get it to work by adding `Pretty (UF.UnionFind meta (TermMeta lang meta))` or similar in lots of places, not just in the `Constraints` file

File: src/Pirouette/SMT/Constraints.hs [hunk 6]
```diff
@@ -137,14 +132,16 @@ conjunct c cs0 =
...
-      -- Finally, we have a /new/ @v@
-      | otherwise = Just $ unifyNewMetaWith cs v u
+    unifyMetaWith cs v u =
+      -- FIXME: We are dropping the optimised union-find here, by using @snd@.
+      case snd $ UF.lookup v (csAssignments cs) of
+        -- @v@ is known to be a given term that has to be unfifiable with @u@
+        Just (Just t) -> unifyWith cs t u
+        -- @v@ is known but does not link to any value. FIXME: Therefore, it
+        -- looks like we are violating an invariant of `unifyNewMetaWith`.
+        Just Nothing -> Just $ unifyNewMetaWith cs v u
```
[the remark this answers is not in the document]
Comment (lucaspena) [reply src/Pirouette/SMT/Constraints.hs:125]: Maybe we can just change the name of `unifyNewMetaWith`? Perhaps something like `unifyUnassignedMetaWith`?

File: src/UnionFind.hs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+import Data.Function ((&))
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import Prettyprinter
+
+data UnionFindCell key value
+  = ChildOf key
+  | Ancestor
+  | AncestorWith value
```
[the remark this answers is not in the document]
Comment (lucaspena) [reply src/UnionFind.hs:29]: If I'm understanding correctly, we need this because of the above issue, where we might have variables that are equal but unassigned. In that case, we can't just disallow `Ancestor` can we?

File: src/UnionFind.hs (hunk 3 above)
Comment (lucaspena) [reply src/UnionFind.hs:197]: I think this is just `join` in `Control.Monad`. Unless I'm misunderstanding something, we should probably just use that.

**Review by lucaspena — COMMENTED** [review]
Some additional comments:

 - The OHearn test has caused issues before, but it makes sense that we could have metavariables potentially equal that are unassigned. But maybe we need some more complex tests if there is generally only this one test where these edge cases come into play.
 - On a similar note, I mentioned before but we probably would want some tests just testing the union-find implementation on its own, especially once the optimizations are implemented.
 - The optimizations would be nice, and it would also be nice to see how much the optimizations affect things. For example, I don't really have any intuition as to how often we do nontrivial unions of metavariables or anything
 - It seems like using a State monad would make things a bit nicer, and apparently maybe even faster!
 - I don't think I understand what persistent union-find means. Is the idea that we would no longer use a tree-like structure to keep track of the equivalences?

Overall, this seems very nice, good work! I suppose if this ultimately gets merged (or rather if Victor's PR gets merged), the most pressing issues would be adding optimizations and adding tests.

File: src/UnionFind.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind.hs:29]: That is correct.

File: src/UnionFind.hs (hunk 3 above)
Comment (Niols) [reply src/UnionFind.hs:197]: This is definitely `join` in `Control.Monad`. For some reason, it didn't show up in Hoogle; I must've done something fishy.

File: src/UnionFind.hs (hunk 4 above)
Comment (Niols) [reply src/UnionFind.hs:78]: I suppose it would just give us a slightly stronger check that the invariants hold. Here, line 80, because there is a `ChildOf key'`, we know that `key'` *must be* in the map. Therefore, we know that `find key' unionFind` will evaluate to `Just`. With the current implementation, though, it could evaluate to `Nothing` without us ever noticing.

File: src/Pirouette/SMT/Constraints.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/Pirouette/SMT/Constraints.hs]: Oh, nice! If you have a version that works, feel free to push it here! I would have expected `Pretty meta` and `Pretty (TermMeta lang meta)` to be exactly the constraints necessary to call `pretty` on a `UnionFind meta (TermMeta lang meta)`, but apparently not, and I don't understand why not.

**Comment by Niols** [comment]
> * The OHearn test has caused issues before, but it makes sense that we could have metavariables potentially equal that are unassigned. But maybe we need some more complex tests if there is generally only this one test where these edge cases come into play.

It is definitely weird that only one test changes behaviour depending on whether or not we allow meta-variables to be equal without assignment. We really need more tests that expose that kind of edge cases, indeed.

> * On a similar note, I mentioned before but we probably would want some tests just testing the union-find implementation on its own, especially once the optimizations are implemented.

Definitely! It will also be good practice for me to have a look at how tests are done. I will try adding tests for the union-find implementation and also maybe some extra tests for the symbolic engine in general (which then would probably escape the scope of this PR).

> * The optimizations would be nice, and it would also be nice to see how much the optimizations affect things. For example, I don't really have any intuition as to how often we do nontrivial unions of metavariables or anything

Well, the weird thing is that, by mimicking the previous implementation, I only ended up using `lookup`, `trivialUnion` and `trivialInsert`, while the whole purpose of the union-find is to have this “fancy” union which takes care of things for you. Basically, I think it shows how deeply-inlined the union-find was, and I think it calls for a full rewrite of `unifyMetaWith` and `unifyNewMetaWith` into (almost only) one single call to `union` with the right merge function.

To be honest, I am not even sure at this point that I understand how this algorithm can possibly work without ever triggering a non-trivial union. I suppose there is an important property of meta-variables that makes this happen (or rather not happen), but I don't see it yet.

> * It seems like using a State monad would make things a bit nicer, and apparently maybe even faster!

It would definitely make the interface of the `UnionFind` module look nicer. And also maybe it would make easier, when using `UnionFind`, to propagate the modifications (and therefore the optimisations, when we implement them)?

> * I don't think I understand what persistent union-find means. Is the idea that we would no longer use a tree-like structure to keep track of the equivalences?

This question was the occasion for me to read on the difference between “persistent” and “immutable”. In short:
- *Persistence* is a property of the *interface*: updates (eg. with `Data.Map.insert`) give you a new pointer to an updated structure but old pointers are still valid. Observations (eg. with `Data.Map.lookup`) cannot distinguish the state of an old pointer before or after updates.
- *Immutability* is an implementation technique where you don't allow rewriting pointers. In particular, it trivially gives you persistence of your data structures.

(If you already knew all that, sorry, then it was only for me :p)
My point was therefore that, as of now, `UnionFind` provides an immutable data-structure. But union-find is a funny case where it is easy to make `find` have a side-effect instead of returning an updated structure while keeping the persistent interface. You get the nice properties of persistence but the interface becomes slightly simpler (because `find` doesn't need to return a union-find anymore). Additionally, you get the speed-up of the optimisation *and this speed-up is shared between diverging versions of the union-find*, ie. if you start from a union-find `u` and update it to `v`, then start from `u` again and update it to `w`, and then you call `find` on `v`, you gain speed-ups for further calls to `find` on `v` *but also on `w`*. In our case where we keep branching at each `if-then-else`, it might actually give us an extra speed-up. This is a consideration for muuuuch later, and it is also out of my Haskell skills because I basically have no idea how to use/implement mutable structures :p

> Overall, this seems very nice, good work! I suppose if this ultimately gets merged (or rather if Victor's PR gets merged), the most pressing issues would be adding optimizations and adding tests.

I'll start by adding tests because it is also very good practice for me. This will allow us to measure the optimisations if we ever decide to implement them.

File: src/UnionFind.hs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+-- | @find key unionFind@ finds the ancestor and value associated to @key@ in
+-- the @unionFind@ structure. It returns a new union-find structure with the
+-- same semantics but optimised for future calls, the ancestor for @key@ and the
+-- found value, @Just Nothing@ if the @key@ is in the structure but without
+-- binding, or @Nothing@ if @key@ is not in the structure.
+--
+-- Prefer using @lookup@.
+--
+-- FIXME: Maybe we prefer a custom type instead of `Maybe (Maybe value)`?
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind.hs:68]: Actually, now that I think about it, I'm not sure it makes sense to return a `Maybe (Maybe value)`: a union-find is supposed to represent a partial map from equivalence classes over a given type to some values. The fact that we don't actually store all the equivalence classes in the map is more an implementation detail: things that are not mentioned in the union-find do exist in the type, they belong to an equivalence class that is a singleton and that isn't mapping to any value. I don't really see why we would want to expose on the outside the difference, and maybe `lookup` should just return a `Maybe value`.

Currently, though, the code in `Pirouette.SMT.Constraints` does use this information which, in my opinion, only shows that it still contains some of the union-find's logic when it shouldn't. I will try to see if I can further rewrite `unify*With` to get rid of these considerations, hopefully improving the readability drastically as well.

And if it doesn't work, I will come back to a custom type, something like
```haskell
data LookupResult value
    = Unknown
    | Unbound         -- or @Known@?
    | BoundTo value   -- or @MapsTo@?
```
or something.

File: src/Pirouette/SMT/Constraints.hs [hunk 9]
```diff
@@ -182,25 +172,27 @@ conjunct c cs0 =
...
 constraintSetToSExpr ::
   forall lang meta m.
-  (LanguageSMT lang, ToSMT meta, PirouetteReadDefs lang m) =>
+  (Ord meta, LanguageSMT lang, ToSMT meta, PirouetteReadDefs lang m) =>
   S.Set Name ->
   ConstraintSet lang meta ->
   m (Bool, UsedAnyUFs, PureSMT.SExpr)
 constraintSetToSExpr knownNames ConstraintSet {..} = do
-  eAssignments <- mapM (runTranslator . uncurry trAssignment) $ M.toList csAssignments
+  -- FIXME: this is dropping the optimised version of the union-find structure.
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/Pirouette/SMT/Constraints.hs:172]: Yes, definitely. Also discussed in https://github.com/tweag/pirouette/pull/150#pullrequestreview-1106236084 and https://github.com/tweag/pirouette/pull/150#issuecomment-1246358175.

File: src/Pirouette/SMT/Constraints.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/Pirouette/SMT/Constraints.hs:125]: This and updating the comment would make sense, at least. But maybe I will get rid of `unify*MetaWith` altogether,
cf https://github.com/tweag/pirouette/pull/150#discussion_r969783049 and https://github.com/tweag/pirouette/pull/150#issuecomment-1246358175.

File: src/UnionFind.hs [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+  value ->
+  UnionFind key value ->
+  UnionFind key value
+trivialInsert = insert (\_ _ -> error "insert was not trivial")
+
+-- | It is guaranteed that the right-hand side of a pair in the list of
+-- equalities is the representant of an equivalence class, and therefore it
+-- occurs on the left-hand side in the list of bindings.
+--
+toLists :: Ord key => UnionFind key value -> (UnionFind key value, [(key, key)], [(key, value)])
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind.hs]: Fixed in cf4f6b4.

File: src/UnionFind.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind.hs]: Note: I merged these two together to simplify other pieces of code. This now looks like:
https://github.com/tweag/pirouette/blob/531cb2e35b7e78e932e9230b8f95e46a98840415/src/UnionFind.hs#L29-L31

File: src/UnionFind.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind.hs]: Actually, the code in `Constraints.hs` did not distinguish between `Nothing` and `Just Nothing` so I stopped making the difference in `find` and `lookup` which now return simply a `Maybe value`.

File: src/Pirouette/SMT/Constraints.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Niols) [reply src/Pirouette/SMT/Constraints.hs]: Commit c92f454 inlines `unifyNewMetaWith` into `unifyMetaWith` because the code had become very short. It also adds a bunch of comments to make things clearer. I actually understood a lot in the process. In particular, why the trivial union and trivial insert worked. In fact, I also understood the whole `unifyWith` function better and why it never needs a non-trivial union or insert. A full blown union-find retrospectively sounds overkill.

File: src/UnionFind.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lucaspena) [reply src/UnionFind.hs]: Seems better to me

File: src/UnionFind.hs (hunk 4 above)
Comment (lucaspena) [reply src/UnionFind.hs:83]: Ah that makes sense. Yea, better guarantees can't hurt

**Comment by lucaspena** [comment]
Thanks for the response, I understand the persistence property now. As for tests, if you would like to pair program some at some point this week or next, let me know. It would be a good learning experience for me as well.

File: src/Pirouette/SMT/Constraints.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucaspena) [reply src/Pirouette/SMT/Constraints.hs]: I pushed up a version that works, though as I mentioned it kind of copies the pretty code in the `UnionFind` module. If it was unclear, I only needed to add these extra constraints if I was trying to use the pretty instance from `UnionFind` instead of copying it, and I don't know why that is. I'll do some more tinkering tomorrow now that some code changed to see if I can figure it out

**Comment by Niols** [comment]
> As for tests, if you would like to pair program some at some point this week or next, let me know. It would be a good learning experience for me as well.

That sounds great; let's do that!

File: src/UnionFind.hs (hunk 4 above)
Comment (Niols) [reply src/UnionFind.hs]: Done in a70eb9d.

Comment (Niols) [reply src/UnionFind.hs]: And this was apparently a good call because the invariant *does* get violated in some tests.

**Review by 0xd34df00d — APPROVED** [review]
That's a really good separation of responsibilities, thanks! In the long run, it'd be cool to have that as a separate library on Hackage, as I can see it useful elsewhere as well, and other libraries have some downsides (at least from a quick survey). Perhaps one could achieve decent performance with underlying mutability with something like `vector`'s/etc `freeze`/`thaw` (did somebody say "linear haskell" too?).

Having said that, I only have minor concerns. In particular, I'm not sure that `Ancestor` should really contain `Maybe` — if the UF user has a use case with partiality, they can just have `UnionFind key (Maybe val)`. At least, I don't feel like `Nothing` is intrinsic to the UF itself. I see there was some discussion about that already, but I'm not sure how relevant it is now, so we can catch up on that separately.

Also, ideally it'd be great to have more tests, especially property-based ones, but that can wait until when (and if) the UF implementation will be factored out to a separate library. Passing the usual pirouette tests is good enough for now, since they exercise enough of UF machinery.

**Comment by aspiwack** [comment]
May I inquire why this PR is still ongoing? It's been open for a month, it should have been merged by now.

File: src/UnionFind/Monad.hs [hunk 11]
```diff
@@ -0,0 +1,62 @@
+module UnionFind.Monad where
+
+import qualified Control.Monad.Trans.State.Lazy as S
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind/Monad.hs]: I'm just not used to having to think about Lazy vs. Strict. Changed in 07329be.

File: src/UnionFind/Monad.hs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+trivialUnion = union (\_ _ -> error "union was not trivial")
+
+insert ::
+  (Ord key, Monad m) =>
+  (value -> value -> value) ->
+  key ->
+  value ->
+  WithUnionFindT key value m ()
+insert merge key value =
+  S.modify $ UF.insert merge key value
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind/Monad.hs]: Changed in 07329be.

File: src/UnionFind/Internal.hs [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+  UnionFind key value ->
+  (UnionFind key value, key, Maybe value)
+find key unionFind =
+  case Map.lookup key $ getMap unionFind of
+    Nothing -> (unionFind, key, Nothing)
+    Just (ChildOf key') ->
+      -- We use @findExn@ because invariant (ii) guarantees that it will
+      -- succeed. If it ever raises an exception, then we found a bug.
+      findExn key' unionFind
+    Just (Ancestor maybeValue) -> (unionFind, key, maybeValue)
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind/Internal.hs:87]: I used to make the distinction between these two cases, but I wasn't sure that it was a good idea. I then realised the client code (that is the implementation of `unifyWith`) never used the distinction and I removed it. I am fine reintroducing it if you think it is useful (maybe in order to be more generic than for Pirouette?)

File: src/UnionFind/Internal.hs [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+      (unionFind2, ancestor2, maybeValue2) = find key2 unionFind1
+   in if ancestor1 == ancestor2
+        then return unionFind2
+        else
+          let onUnionFind2sMap update = return $ UnionFind $ update $ getMap unionFind2
+           in case (maybeValue1, maybeValue2) of
+                (Nothing, Nothing) ->
+                  onUnionFind2sMap $
+                    Map.insert key1 (ChildOf key2)
+                      . Map.insert key2 (Ancestor Nothing)
```
[the remark this answers is not in the document]
Comment (Niols) [reply src/UnionFind/Internal.hs:169]: `Ancestor` can contain a `Nothing` if you `union` two keys that were unknown until now. That would lead you exactly line 166. However, now that you point me to this function, it looks potentially buggy to me. I think it goes to an inconsistent state if we merge two equivalence classes bound to no value together. As of now, I think it uses the information that `maybeValuen == Nothing` to deduce that `keyn` is unknown to the structure, but that is simply not true. I think it can get fixed by replacing all uses of `keyn` by `ancestorn`, but I need to think about it some more. We should also have tests to catch that sort of mistakes.

**Comment by Niols** [comment]
@0xd34df00d 

> That's a really good separation of responsibilities, thanks! In the long run, it'd be cool to have that as a separate library on Hackage, as I can see it useful elsewhere as well, and other libraries have some downsides (at least from a quick survey).

I had the same impression about other libraries which is why I decided to go for a home implementation. I suppose it could be separated from Pirouette indeed.
 
> Having said that, I only have minor concerns. In particular, I'm not sure that `Ancestor` should really contain `Maybe` — if the UF user has a use case with partiality, they can just have `UnionFind key (Maybe val)`. At least, I don't feel like `Nothing` is intrinsic to the UF itself. I see there was some discussion about that already, but I'm not sure how relevant it is now, so we can catch up on that separately.

It is a good question and I'm not sure what the right answer is. For now, `Nothing` is used to convey the fact that we know of an equivalence class but we haven't bound a value to it. It seems sometimes important, for instance for code that learns equivalences and assignments one by one and not necessarily in an order that would work without a `Maybe`.

The code would probably be nicer without a `Maybe` under `Ancestor`. At least, it would avoid merging two different pieces of information into one, and it would avoid potential bugs like the one discussed in https://github.com/tweag/pirouette/pull/150#discussion_r991250090.

> Also, ideally it'd be great to have more tests, especially property-based ones, but that can wait until when (and if) the UF implementation will be factored out to a separate library. Passing the usual pirouette tests is good enough for now, since they exercise enough of UF machinery.

I would be all for property-based tests for this little piece of code. I hope that the usual Pirouette tests are good enough but I am not sure. Do we have property-based testing machinery in the Pirouette project or would that be adding big dependencies?

**Comment by Niols** [comment]
@aspiwack 

> May I inquire why this PR is still ongoing? It's been open for a month, it should have been merged by now.

You are completely right, this PR should be merged by now (or closed; but not stalled). I left it hanging because I wasn't sure that it was the right way to go. I will make sure that we review and merge it quickly.

**Comment by Niols** [comment]
This PR seems ready to be merged. It passes all the tests that were previously there as well as the newly-introduced ones. It now implements the classic optimisations of the Union-Find structure. It could be further optimised by using a structure different from a Map (but still persistent), but I don't think it makes much sense to go through the trouble right now. We can always iterate later on if we feel the need.
